server/src/main/java/org/elasticsearch/health/HealthIndicatorServiceBase.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.health;
+
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.coordination.NoMasterBlockService;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.gateway.GatewayService;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.util.Collections;
+
+public abstract class HealthIndicatorServiceBase implements HealthIndicatorService {
+
+    static final String COULD_NOT_DETERMINE_HEALTH = "Health could not be determined. Try executing health api again.";
+
+    private final ClusterService clusterService;
+
+    public HealthIndicatorServiceBase(ClusterService clusterService) {
+        this.clusterService = clusterService;
+    }
+
+    @Override
+    public HealthIndicatorResult calculate(boolean calculateDetails) {
+        ClusterState clusterState = clusterService.state();
+        if (stateIsKnown(clusterState)) {
+            return doCalculate(clusterState, calculateDetails);
+        } else {
+            return doCalculateUnknown(clusterState, calculateDetails);
+        }
+    }
+
+    protected abstract HealthIndicatorResult doCalculate(ClusterState clusterState, boolean calculateDetails);
+
+    protected HealthIndicatorResult doCalculateUnknown(ClusterState clusterState, boolean calculateDetails) {
```

Comment:
I believe we need another approach here. 

We have 2 use cases I believe:

1. This coordinating node has a block on the cluster state (which is what is currently tackled in this PR)

2. The indicator responsible with checking the master node (currently `instance-has-master` but soon it'll be the `master-is-stable` indicator) returned RED so some of the other indicators should return `UNKNOWN`. I believe we should have this ability to say "indicator 1 is RED so indicator 2,3,4 must be UNKNOWN now" (currently _all_ the other indicators should return `UNKNOWN` if `instance-has-master` returned RED as they rely on the cluster state, however in the future maybe the `network` indicator will be able to report some health information about this node's connections or such)

Reply:
Are we sure we want to return an UNKNOWN in the case that master is not stable? If the master is unstable but we _are_ connected to a master node at this moment then it stands to reason that the other indicators should be able to produce their results right?

Reply:
That's a good question. The `master-is-stable` name is maybe not the best name for what it does. But it's a superset of `instance-has-master` and I believe in the case it returns `RED` ie. we don't have a master or we can't join a master, or even `YELLOW` (as that indicates "well, the master node is flapping and the cluster isn't reliably working") I don't think we can rely on the local cluster state to surely say there are no unassigned shards for e.g.

I believe we need a model that check both the cluster state blocks (like we have in this PR) but also a model that says "oh indicator X is RED, these other ones need to be UNKNOWN now"

I believe with the current implementation we can't express that. If we have _some_ cluster state without any blocks I can have `instance-has-master` returning RED (ie. no master in the cluster state) and at the same time `shards_availability` GREEN (which would be incorrect to indicate - this should be UNKNOWN)